crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2328,7 +2343,9 @@ impl<'db> TypeInferenceBuilder<'db> {
         } = for_statement;
 
         self.infer_target(target, iter, |db, iter_ty| {
-            iter_ty.iterate(db).unwrap_without_diagnostic()
+            // It's okay to ignore the `iterate` outcome here because
+            // `infer_for_statement_definition` reports a diagnostic if `iter_ty` isn't iterable.
```

Comment:
I think this isn't quite right, and there should be a TODO here (though a low-priority one). If the `for` statement is not a definition, because the target is not a `Name` (Python allows e.g. `for x.attr in ...` and `for x[i] in ...`, strange as they look), we will never call `infer_for_statement_definition`, and we should ideally check the iterable for iterability and emit a diagnostic if it's not iterable.